Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -271,54 +264,214 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
+        completeTokenWaiters(passing: nil)
     }
-    
-    // TODO: Not used & tested yet -> CIS-224
-    func refreshToken(completion: @escaping (Error?) -> Void) {
-        log.assert(config.tokenProvider != nil, "You can't call `refreshToken` when `Config.tokenProvider` is nil.")
-        
-        config.tokenProvider?(config.apiKey, currentUserId, { [weak self] (token) in
-            guard let token = token else {
-                completion(ClientError.MissingToken("Can't connect because `TokenProvider` didn't return a valid token."))
-                return
-            }
-            self?.currentToken = token
-            completion(nil)
-        })
-    }
-    
+}
+
+private extension _ChatClient {
     func createBackgroundWorkers() {
         backgroundWorkers = workerBuilders.map { builder in
             builder(self.databaseContainer, self.apiClient)
         } + eventWorkerBuilders.map { builder in
             builder(self.databaseContainer, self.eventNotificationCenter, self.apiClient)
         }
     }
-    
+
+    func completeConnectionIdWaiters(passing connectionId: String?) {
+        _connectionIdWaiters.mutate { waiters in
+            waiters.forEach { $0(connectionId) }
+            waiters.removeAll()
+        }
+    }
+
+    func completeTokenWaiters(passing token: Token?) {
+        _tokenWaiters.mutate { waiters in
+            waiters.forEach { $0(token) }
+            waiters.removeAll()
+        }
+    }
+
     /// Before CIS-555, `webSocketClient` was initialized in `createBackgroundWorkers` in `init` since it was passed
     /// to the background workers. Now, background workers do not access `webSocketClient` and it's not being
     /// initialized in `init`, so we initialize it here explicitly.
-    private func createWebSocketClient() {
+    func createWebSocketClient() {
         _ = webSocketClient
     }
+
+    func prepareEnvironment(newToken: Token, completion: @escaping (Error?) -> Void) {
+        // Update web-socket endpoint.
+        webSocketClient?.connectEndpoint = .webSocketConnect(userId: newToken.userId)
+
+        // Check token is for current user.
+        guard newToken.userId != currentUserId else {
+            // Update the current token with the new one.
+            currentToken = newToken
+
+            // It makes more sense to create background workers here
+            // other than in `init` because workers without currently logged-in
+            // user do nothing.
+            if backgroundWorkers?.isEmpty ?? true {
+                createBackgroundWorkers()
+            }
+
+            completion(nil)
+            return
+        }
+
+        // Setting a new user is not possible in connectionless mode.
+        guard config.isClientInActiveMode else {
+            completion(ClientError.ClientIsNotInActiveMode())
+            return
+        }
+
+        // Cancel all API requests since they are related to the previous user.
+        completeTokenWaiters(passing: nil)
```

Comment:
This needs to be done for passive client, too

Reply:
Indeed! 👍 

Will also add forwarding the updated token. Someone might call the API right after the client is instantiated and the request will be waiting for the token:
```swift
guard newToken.userId != currentUserId else {
 ...
 // Forward the new token to waiting requests.
 completeTokenWaiters(passing: newToken)
 ...
}
```

Reply:
Great 👏